### PR TITLE
Test Shared Subscription

### PR DESCRIPTION
**Pull request by alfred2g**
Shared Subscription test

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/Mqtt5ClientTest.cpp [hunk 1]
```diff
@@ -1936,6 +1939,234 @@ static int s_TestMqtt5WillTest(Aws::Crt::Allocator *allocator, void *)
...
+            int message_int = atoi(message_string.c_str());
+            ASSERT_TRUE(message_int < MESSAGE_NUMBER);
+            ++receivedMessages[message_int];
+            client1_received = true;
+
+            bool exchanged = false;
+            int desired = 11;
+            int tested = 10;
+            client_messages++;
+            exchanged = client_messages.compare_exchange_strong(tested, desired);
```
Comment (sfod) [inline tests/Mqtt5ClientTest.cpp:1996]: Trivial: Callbacks will be called on a single thread (the eventloop thread), so no data race should be possible. It's not a big deal, but I'd simplify it for maintainability.
I think, it's safe to make `client_messages` non-atomic.

File: tests/Mqtt5ClientTest.cpp [hunk 2]
```diff
@@ -1936,6 +1939,234 @@ static int s_TestMqtt5WillTest(Aws::Crt::Allocator *allocator, void *)
...
+            if (exchanged == true)
+            {
+                client_received.set_value();
+            }
+        }
+        return 0;
```
Comment (sfod) [inline tests/Mqtt5ClientTest.cpp:2035]: Trivial nit: the callbacks for both subscribers are basically the same, so it's worth introducing a standalone function/lambda.

File: tests/Mqtt5ClientTest.cpp [hunk 3]
```diff
@@ -1936,6 +1939,234 @@ static int s_TestMqtt5WillTest(Aws::Crt::Allocator *allocator, void *)
...
+    ASSERT_TRUE(mqtt5Client->Subscribe(subscribe, onSubAck));
+    suback.get_future().wait();
+
+    suback = std::promise<void>();
+
+    /* subscribe second client */
+    ASSERT_TRUE(mqtt5Client2->Subscribe(subscribe2, onSubAck));
+    suback.get_future().wait();
+
+    /* Publish message 10 to test topic */
```
Comment (sfod) [inline tests/Mqtt5ClientTest.cpp:2114]: Trivial: Publish 10 messages to test topic

File: tests/Mqtt5ClientTest.cpp [hunk 4]
```diff
@@ -1936,6 +1939,234 @@ static int s_TestMqtt5WillTest(Aws::Crt::Allocator *allocator, void *)
...
+    unsubList.push_back(TEST_TOPIC);
+    std::shared_ptr<Mqtt5::UnsubscribePacket> unsubscribe_client1 =
+        std::make_shared<Mqtt5::UnsubscribePacket>(allocator);
+    unsubscribe_client1->WithTopicFilters(unsubList);
+    ASSERT_TRUE(mqtt5Client->Unsubscribe(unsubscribe_client1));
+
+    std::shared_ptr<Mqtt5::UnsubscribePacket> unsubscribe_client2 =
+        std::make_shared<Mqtt5::UnsubscribePacket>(allocator);
+    unsubscribe_client2->WithTopicFilters(unsubList);
+    ASSERT_TRUE(mqtt5Client->Unsubscribe(unsubscribe_client2));
```
Comment (sfod) [inline tests/Mqtt5ClientTest.cpp:2137]: Should be `mqtt5Client2`

**Review by sfod — APPROVED** [review]

File: tests/Mqtt5ClientTest.cpp (hunk 1 above)
Comment (sfod) [reply tests/Mqtt5ClientTest.cpp:1996]: Nope, I'm wrong, so nevermind

Comment (alfred2g) [reply tests/Mqtt5ClientTest.cpp:1996]: as discussed off line, threads might come on different threads as well...

**Pull request merged**